File: pg/BO/ssap-partner-portal-bo-frontend/src/models/ManHour.ts
import { format } from 'date-fns'

import { getCurrentDate } from '~/utils/common-functions'
import {
  IGetSummaryManHourTypeResponse,
  IManHour,
  IGetSummaryManHourTypeRequest,
  IGetSummaryServiceTypesManHoursRequest,
  IGetSummaryServiceTypesManHoursResponse,
  IGetSummarySupporterOrganizationsManHoursRequest,
  IGetSummarySupporterOrganizationsManHourResponse,
  IProject,
  IGetSummaryProjectManHourAlertsRequest,
  IGetSupporterOrganizationsRequest,
  ISummaryManHour,
  IContractTime,
  ISummarySupporterManHour,
  IGetSummarySupporterManHoursRequest,
  IGetSummarySupporterManHoursResponse,
  IServiceTypeItem,
  IGetServiceTypesRequest,
  ISupporterOrganization,
  ISummaryManHourTypeHeader,
  IgetSummaryProjectManHourAlertsResponse,
  ISummaryProjectManHourAlertsProjectItems,
  IGetSummaryProjectManHourAlertDetailResponse,
  IManHours,
  IDirectSupportManHoursItem,
  IDirectSupportManHours,
  IPreSupportManHoursItem,
  IPreSupportManHours,
  ISalesSupportManHoursItem,
  ISalesSupportManHours,
  ISsapManHours,
  IHolidaysManHours,
  ISummarySupportManHours,
  IGetManHourBySupporterUserIdResponse,
  IGetManHourBySupporterUserIdRequest,
  IUpdateManHourRequest,
  IUpdateManHourResponse,
  IUpdateManHourRequestParams,
  IUpdateManHourBySupporterUserIdRequest,
} from '@/types/ManHour'
import { Api } from '~/plugins/api'

const $api = new Api()

export class SummaryManHourTypeHeader implements ISummaryManHourTypeHeader {
  supporterOrganizationManHours = []
  supporterOrganizationTotal = []
}
export class ManHourType implements IManHour {
  manHourTypeName = ''
  subName = ''
  serviceTypeName = ''
  roleName = ''
  supporterOrganizationTotal = []
  supporterOrganizationManHours = []
}

export class GetSummaryManHourTypeResponse
  implements IGetSummaryManHourTypeResponse
{
  yearMonth = ''
  header = [] as ISummaryManHourTypeHeader[]
  manHours = [] as ManHourType[]
}

export class Project implements IProject {
  projectId = ''
  projectName = ''
  customerId = ''
  contractType = ''
  thisMonthDirectSupportManHourMain = 0
  thisMonthDirectSupportManHourSub = 0
  thisMonthPreSupportManHour = 0
  thisMonthContractTime = 0
  totalProcessYPercent = 0
}

export class GetSummaryServiceTypesManHoursRequest
  implements IGetSummaryServiceTypesManHoursRequest
{
  year = parseInt(format(getCurrentDate(), 'yyyy'))
  month = parseInt(format(getCurrentDate(), 'MM'))
}

export class GetSummaryServiceTypesManHoursResponse
  implements IGetSummaryServiceTypesManHoursResponse
{
  serviceTypeId = ''
  serviceTypeName = ''
  directSupportManHourFactor = 0
  projects = [] as Project[]
}

export async function GetSummaryServiceTypesManHours(
  request: GetSummaryServiceTypesManHoursRequest
) {
  return await $api.get<GetSummaryServiceTypesManHoursResponse[]>(
    `man-hours/summary/service-types`,
    request
  )
}

export class GetSummarySupporterOrganizationsManHoursRequest
  implements IGetSummarySupporterOrganizationsManHoursRequest
{
  year = parseInt(format(getCurrentDate(), 'yyyy'))
  month = parseInt(format(getCurrentDate(), 'MM'))
  supporterOrganizationId?: string
}

export class GetSummarySupporterOrganizationsManHourResponse
  implements IGetSummarySupporterOrganizationsManHourResponse
{
  supporterOrganizationId = ''
  supporterOrganizationName = ''
  annualSales = 0
  monthlySales = 0
  monthlyProjectPrice = 0
  monthlyContractTime = 0
  monthlyWorkTime = 0
  monthlyWorkTimeRate = 0
  monthlySupporters = 0
  monthlyManHour = 0
  monthlyOccupancyRate = 0
  monthlyOccupancyTotalTime = 0
  monthlyOccupancyTotalRate = 0
  updateAt = ''
}

export async function GetSummarySupporterOrganizationsManHours(
  request: GetSummarySupporterOrganizationsManHoursRequest
) {
  return await $api.get<GetSummarySupporterOrganizationsManHourResponse[]>(
    `man-hours/summary/supporter-organizations`,
    request
  )
}

export class GetSummaryManHourTypeRequest
  implements IGetSummaryManHourTypeRequest
{
  year = parseInt(format(getCurrentDate(), 'yyyy'))
  month = parseInt(format(getCurrentDate(), 'MM'))
}

export async function GetSummaryManHourType(
  request: GetSummaryManHourTypeRequest
) {
  return await $api.get<GetSummaryManHourTypeResponse>(
    `/man-hours/summary/type`,
    request
  )
}

export class GetSummaryProjectManHourAlertsRequest
  implements IGetSummaryProjectManHourAlertsRequest
{
  public year = 0
  public month = 0
  public supporterOrganizationId? = ''
  public serviceTypeId? = ''
}

export async function GetSummaryProjectManHourAlerts(
  request: GetSummaryProjectManHourAlertsRequest
) {
  return await $api.get<GetSummaryProjectManHourAlertsResponse>(
    `/man-hours/summary/project-man-hour-alerts`,
    request
  )
}

export class SummaryManHour implements ISummaryManHour {
  public direct = 0
  public pre = 0
  public sales = 0
  public ssap = 0
  public others = 0
  public total = 0
}

export class ContractTime implements IContractTime {
  public producer = 0
  public accelerator = 0
  public total = 0
}

export class SummarySupporterManHour implements ISummarySupporterManHour {
  public supporterOrganizationId = ''
  public supporterOrganizationName = ''
  public supporterId = ''
  public supporterName = ''
  public summaryManHour = [] as SummaryManHour[]
  public contractTime = [] as ContractTime[]
  public isConfirm = true
}

export class GetSummarySupporterManHoursRequest
  implements IGetSummarySupporterManHoursRequest
{
  year = parseInt(format(getCurrentDate(), 'yyyy'))
  month = parseInt(format(getCurrentDate(), 'MM'))
}

export class GetSummarySupporterManHoursResponse
  implements IGetSummarySupporterManHoursResponse
{
  yearMonth = ''
  manHours = [] as SummarySupporterManHour[]
}

export async function GetSummarySupporterManHours(
  request: GetSummarySupporterManHoursRequest
) {
  return await $api.get<GetSummarySupporterManHoursResponse>(
    `/man-hours/summary/supporter`,
    request
  )
}

//本来Masterから取得
export class SupporterOrganization implements ISupporterOrganization {
  public id = ''
  public name = ''
  public shortName = ''
}

export class GetSupporterOrganizationsRequest
  implements IGetSupporterOrganizationsRequest
{
  public year = parseInt(format(getCurrentDate(), 'yyyy'))
  public month = parseInt(format(getCurrentDate(), 'MM'))
}

export class GetSupporterOrganizationsResponse extends Array<SupporterOrganization> {}

export async function GetSupporterOrganizations(
  request: GetSupporterOrganizationsRequest
) {
  return await $api.get<GetSupporterOrganizationsResponse>(
    `/masters/supporter-organizations`,
    request
  )
}
export class ServiceTypeItem implements IServiceTypeItem {
  public id = ''
  public name = ''
}

export class ServiceTypeRequest implements IGetServiceTypesRequest {
  public year = parseInt(format(getCurrentDate(), 'yyyy'))
  public month = parseInt(format(getCurrentDate(), 'MM'))
}

export class GetServiceTypesResponse extends Array<ServiceTypeItem> {}

export async function GetServiceTypes(request: IGetServiceTypesRequest) {
  return await $api.get<GetServiceTypesResponse>(
    `/masters/service-types`,
    request
  )
}

export class GetSummaryProjectManHourAlertsResponse
  implements IgetSummaryProjectManHourAlertsResponse
{
  summaryThisMonthContractTime = 0
  projects = [] as ISummaryProjectManHourAlertsProjectItems[]
}

export class ManHourResponse implements IManHours {
  supporterUserId = ''
  supporterUserName = ''
  role = ''
  thisMonthSupporterDirectSupportManHours = 0
  thisMonthSupporterPreSupportManHours = 0
  summarySupporterDirectSupportManHours = 0
  summarySupporterPreSupportManHours = 0
}

export class GetSummaryProjectManHourAlertDetailResponse
  implements IGetSummaryProjectManHourAlertDetailResponse
{
  projectId = ''
  projectName = ''
  customerId = ''
  customerName = ''
  manHours: ManHourResponse[] = []
}

export class GetSummaryProjectManHourAlertDetailRequest {
  public year: number = 0
  public month: number = 0
  public projectId: string = ''
}

export async function getSummaryProjectManHourAlertDetail(
  request: GetSummaryProjectManHourAlertDetailRequest
) {
  return await $api.get<GetSummaryProjectManHourAlertDetailResponse>(
    `/man-hours/summary/project-man-hour-alert`,
    request
  )
}

export class DirectSupportManHoursItem implements IDirectSupportManHoursItem {
  public projectId = ''
  public projectName = ''
  public role = ''
  public serviceType = ''
  public kartenManHour = 0
  public inputManHour = 0
}
export class DirectSupportManHours implements IDirectSupportManHours {
  public items = [] as DirectSupportManHoursItem[]
  public memo = ''
}
export class PreSupportManHoursItem implements IPreSupportManHoursItem {
  public projectId = ''
  public projectName = ''
  public role = ''
  public serviceType = ''
  public inputManHour = 0
}
export class PreSupportManHours implements IPreSupportManHours {
  public items = [] as PreSupportManHoursItem[]
  public memo = ''
}

export class SalesSupportManHoursItem implements ISalesSupportManHoursItem {
  public projectName = ''
  public customerId = ''
  public customerName = ''
  public type = ''
  public inputManHour = 0
}
export class SalesSupportManHours implements ISalesSupportManHours {
  public items = [] as SalesSupportManHoursItem[]
  public memo = ''
}

export class SsapManHours implements ISsapManHours {
  public meeting = 0
  public study = 0
  public learning = 0
  public newService = 0
  public startdash = 0
  public improvement = 0
  public ssap = 0
  public qc = 0
  public accounting = 0
  public management = 0
  public officeWork = 0
  public others = 0
  public memo = ''
}

export class HolidaysManHours implements IHolidaysManHours {
  public paidHoliday = 0
  public holiday = 0
  public private = 0
  public others = 0
  public departmentOthers = 0
  public memo = ''
}

export class SummarySupportManHours implements ISummarySupportManHours {
  public direct = 0
  public pre = 0
  public sales = 0
  public ssap = 0
  public others = 0
  public total = 0
}

export class GetManHourBySupporterUserIdResponse
  implements IGetManHourBySupporterUserIdResponse
{
  public yearMonth = ''
  public supporterUserId = ''
  public supporterUserName = ''
  public supporterOrganizationId = ''
  public supporterOrganizationName = ''
  public directSupportManHours = [] as DirectSupportManHours[]
  public preSupportManHours = [] as PreSupportManHours[]
  public salesSupportManHours = [] as SalesSupportManHours[]
  public ssapManHours = {} as SsapManHours
  public holidaysManHours = {} as HolidaysManHours
  public summarySupportManHours = {} as SummarySupportManHours
  public isConfirm = true
  public createId = ''
  public createUserName = ''
  public createAt = ''
  public updateId = ''
  public updateUserName = ''
  public updateAt = ''
  public version = 0
}

export class GetManHourBySupporterUserIdRequest
  implements IGetManHourBySupporterUserIdRequest
{
  public year = parseInt(format(getCurrentDate(), 'yyyy'))
  public month = parseInt(format(getCurrentDate(), 'MM'))
}

export async function GetManHourBySupporterUserId(
  supporterUserId: string,
  request: GetManHourBySupporterUserIdRequest
) {
  return await $api.get<GetManHourBySupporterUserIdResponse>(
    `/man-hours/${supporterUserId}`,
    request
  )
}

export class UpdateManHourRequestParams implements IUpdateManHourRequestParams {
  public id = ''
  public version = 0
  public year = parseInt(format(getCurrentDate(), 'yyyy'))
  public month = parseInt(format(getCurrentDate(), 'MM'))
}

export class UpdateManHourRequest implements IUpdateManHourRequest {
  public supporterUserName = ''
  public supporterOrganizationId = ''
  public supporterOrganizationName = ''
  public directSupportManHours = [] as DirectSupportManHours[]
  public preSupportManHours = [] as PreSupportManHours[]
  public salesSupportManHours = [] as SalesSupportManHours[]
  public ssapManHours = {} as SsapManHours
  public holidaysManHours = {} as HolidaysManHours
  public summarySupportManHours = {} as SummarySupportManHours
  public isConfirm = true
}

export class UpdateManHourResponse implements IUpdateManHourResponse {
  public yearMonth = ''
  public supporterUserId = ''
  public supporterUserName = ''
  public supporterOrganizationId = ''
  public supporterOrganizationName = ''
  public directSupportManHours = [] as DirectSupportManHours[]
  public preSupportManHours = [] as PreSupportManHours[]
  public salesSupportManHours = [] as SalesSupportManHours[]
  public ssapManHours = {} as SsapManHours
  public holidaysManHours = {} as HolidaysManHours
  public summarySupportManHours = {} as SummarySupportManHours
  public isConfirm = true
  public createId = ''
  public createUserName = ''
  public createAt = ''
  public updateId = ''
  public updateUserName = ''
  public updateAt = ''
  public version = 0
}

export class UpdateManHourBySupporterUserIdRequest
  implements IUpdateManHourBySupporterUserIdRequest
{
  public supporterUserName = ''
  public supporterOrganizationId = ''
  public supporterOrganizationName = ''
  public directSupportManHours = [] as DirectSupportManHours[]
  public preSupportManHours = [] as PreSupportManHours[]
  public salesSupportManHours = [] as SalesSupportManHours[]
  public ssapManHours = {} as SsapManHours
  public holidaysManHours = {} as HolidaysManHours
  public isConfirm = true
}

export async function UpdateManHourBySupporterUserId(
  supporterUserId: string,
  version: number,
  year: number,
  month: number,
  request: UpdateManHourBySupporterUserIdRequest
) {
  return await $api.put<UpdateManHourResponse>(
    `/man-hours/${supporterUserId}?version=${version}&year=${year}&month=${month}`,
    request
  )
}

File: BO/ssap-partner-portal-bo-backend/app/routers/man_hour.py
from typing import List

from fastapi import APIRouter, Depends, Path
from fastapi import status as api_status

from app.auth.auth import get_current_user_factory
from app.resources.const import UserRoleType
from app.schemas.man_hour import (
    GetManHourBySupporterUserIdResponse,
    GetSummaryManHourTypeResponse,
    GetSummaryProjectManHourAlertQuery,
    GetSummaryProjectManHourAlertResponse,
    GetSummaryProjectManHourAlertsQuery,
    GetSummaryProjectManHourAlertsResponse,
    GetSummaryServiceTypesManHoursResponse,
    GetSummarySupporterManHoursQuery,
    GetSummarySupporterManHoursResponse,
    GetSummarySupporterOrganizationsManHoursQuery,
    GetSummarySupporterOrganizationsManHoursResponse,
    UpdateManHourBySupporterUserIdQuery,
    UpdateManHourBySupporterUserIdRequest,
    UpdateManHourBySupporterUserIdResponse,
    YaerMonthQuery,
)
from app.service.man_hour_service import ManHourService

router = APIRouter()


@router.get(
    "/man-hours/summary/supporter-organizations",
    tags=["ManHour"],
    description="支援者組織（課）別工数指標を取得します。 月別実績 SSAP粗利シミレーションとキーとなる指標 用",
    response_model=List[GetSummarySupporterOrganizationsManHoursResponse],
    status_code=api_status.HTTP_200_OK,
)
def get_summary_supporter_organizations_man_hours(
    query_params: GetSummarySupporterOrganizationsManHoursQuery = Depends(),
    current_user=Depends(
        get_current_user_factory(
            accessible_roles={
                UserRoleType.SYSTEM_ADMIN.key,
                UserRoleType.SALES_MGR.key,
                UserRoleType.SUPPORTER_MGR.key,
                UserRoleType.MAN_HOUR_OPS.key,
                UserRoleType.BUSINESS_MGR.key,
            }
        )
    ),
) -> List[GetSummarySupporterOrganizationsManHoursResponse]:
    """Get /api/man-hours/summary/supporter-organizations 支援者組織(課)別工数取得API

    Args:
        year (int): 対象年
        month (int): 対象月
        supporterOrganizationId (str): 支援者組織ID
        current_user (Behavior, optional): 認証済みのユーザー

    Returns:
        GetSummarySupporterOrganizationsManHoursResponse: 取得結果
    """
    return ManHourService.get_summary_supporter_organizations_man_hours(
        year=query_params.year,
        month=query_params.month,
        supporter_organization_id=query_params.supporter_organization_id,
        current_user=current_user,
    )


@router.get(
    "/man-hours/summary/service-types",
    tags=["ManHour"],
    description="サービス種別別工数指標を取得します。 月別実績 案件別契約時間と直接寄与工数（対面時間）内訳 描画用",
    response_model=List[GetSummaryServiceTypesManHoursResponse],
    status_code=api_status.HTTP_200_OK,
)
def get_summary_service_types_man_hours(
    query_params: YaerMonthQuery = Depends(),
    current_user=Depends(
        get_current_user_factory(
            accessible_roles={
                UserRoleType.SYSTEM_ADMIN.key,
                UserRoleType.SALES_MGR.key,
                UserRoleType.SUPPORTER_MGR.key,
                UserRoleType.MAN_HOUR_OPS.key,
                UserRoleType.BUSINESS_MGR.key,
            }
        )
    ),
) -> List[GetSummaryServiceTypesManHoursResponse]:
    """Get /api/man-hours/summary/service-types サービス種別別工数指標一覧取得API

    Args:
        year (int): 対象年
        month (int): 対象月
        current_user (Behavior, optional): 認証済みのユーザー

    Returns:
        List[GetSummaryServiceTypesManHoursResponse]: 取得結果
    """

    return ManHourService.get_summary_service_types_man_hours(
        year=query_params.year,
        month=query_params.month,
        current_user=current_user,
    )


@router.get(
    "/man-hours/summary/project-man-hour-alerts",
    tags=["ManHour"],
    description="案件毎の工数状況とアラート一覧を取得します。",
    response_model=GetSummaryProjectManHourAlertsResponse,
    status_code=api_status.HTTP_200_OK,
)
def get_summary_project_man_hour_alerts(
    query_params: GetSummaryProjectManHourAlertsQuery = Depends(),
    current_user=Depends(
        get_current_user_factory(
            accessible_roles={
                UserRoleType.SYSTEM_ADMIN.key,
                UserRoleType.SALES_MGR.key,
                UserRoleType.SUPPORTER_MGR.key,
                UserRoleType.MAN_HOUR_OPS.key,
                UserRoleType.BUSINESS_MGR.key,
            }
        )
    ),
) -> GetSummaryProjectManHourAlertsResponse:
    """Get /api/man-hours/summary/project-man-hour-alerts 案件別工数アラート一覧取得API

    Args:
        year (int): 対象年
        month (int): 対象月
        supporterOrganizationId (str): 支援者組織ID
        serviceTypeId (str): 支援者組織ID
        current_user (Behavior, optional): 認証済みのユーザー

    Returns:
        GetSummaryProjectManHourAlertsResponse: 取得結果
    """

    return ManHourService.get_summary_project_man_hour_alerts(
        year=query_params.year,
        month=query_params.month,
        supporter_organization_id=query_params.supporter_organization_id,
        service_type_id=query_params.service_type_id,
        current_user=current_user,
    )


@router.get(
    "/man-hours/summary/project-man-hour-alert",
    tags=["ManHour"],
    description="案件指定の工数状況とアラートを取得します。",
    response_model=GetSummaryProjectManHourAlertResponse,
    status_code=api_status.HTTP_200_OK,
)
def get_summary_project_man_hour_alert(
    query_params: GetSummaryProjectManHourAlertQuery = Depends(),
    current_user=Depends(
        get_current_user_factory(
            accessible_roles={
                UserRoleType.SYSTEM_ADMIN.key,
                UserRoleType.SALES_MGR.key,
                UserRoleType.SUPPORTER_MGR.key,
                UserRoleType.MAN_HOUR_OPS.key,
                UserRoleType.BUSINESS_MGR.key,
            }
        )
    ),
) -> GetSummaryProjectManHourAlertResponse:
    """Get /api/man-hours/summary/project-man-hour-alert 案件別工数アラート取得API

    Args:
        year (int): 対象年
        month (int): 対象月
        projectId (str): 案件ID
        current_user (Behavior, optional): 認証済みのユーザー

    Returns:
        GetSummaryProjectManHourAlertResponse: 取得結果
    """

    return ManHourService.get_summary_project_man_hour_alert(
        year=query_params.year,
        month=query_params.month,
        project_id=query_params.project_id,
        current_user=current_user,
    )


@router.get(
    "/man-hours/{supporter_user_id}",
    tags=["ManHour"],
    description="支援者単位で支援工数を取得します。",
    response_model=GetManHourBySupporterUserIdResponse,
    status_code=api_status.HTTP_200_OK,
)
def get_man_hour_by_supporter_user_id(
    query_params: YaerMonthQuery = Depends(),
    supporter_user_id: str = Path(..., title="支援者ユーザーID"),
    current_user=Depends(
        get_current_user_factory(
            accessible_roles={
                UserRoleType.SYSTEM_ADMIN.key,
                UserRoleType.SALES_MGR.key,
                UserRoleType.SUPPORTER_MGR.key,
                UserRoleType.MAN_HOUR_OPS.key,
                UserRoleType.BUSINESS_MGR.key,
            }
        )
    ),
) -> GetManHourBySupporterUserIdResponse:
    """Get /api/man-hours/{supporter_user_id} 支援者別工数取得API

    Args:
        year (int): 対象年
        month (int): 対象月
        supporter_user_id (str): 支援者ユーザーID
        current_user (Behavior, optional): 認証済みのユーザー

    Returns:
        GetManHourBySupporterUserIdResponse: 取得結果
    """

    return ManHourService.get_man_hour_by_supporter_user_id(
        year=query_params.year,
        month=query_params.month,
        supporter_user_id=supporter_user_id,
        current_user=current_user,
    )


@router.put(
    "/man-hours/{supporter_user_id}",
    tags=["ManHour"],
    description="支援者単位で支援工数を更新します。",
    response_model=UpdateManHourBySupporterUserIdResponse,
    status_code=api_status.HTTP_200_OK,
)
def update_man_hour_by_supporter_user_id(
    body: UpdateManHourBySupporterUserIdRequest,
    query_params: UpdateManHourBySupporterUserIdQuery = Depends(),
    supporter_user_id: str = Path(..., title="支援者ユーザーID"),
    current_user=Depends(
        get_current_user_factory(
            accessible_roles={
                UserRoleType.SYSTEM_ADMIN.key,
                UserRoleType.MAN_HOUR_OPS.key,
            }
        )
    ),
) -> UpdateManHourBySupporterUserIdResponse:
    """Put /api/man-hours/{supporter_user_id} 支援者別工数更新API

    Args:
        year (int): 対象年
        month (int): 対象月
        version (int): ロックキー（楽観ロック制御）
        supporter_user_id (str): 支援者ユーザーID
        current_user (Behavior, optional): 認証済みのユーザー

    Returns:
        UpdateManHourBySupporterUserIdResponse: 更新結果
    """

    return ManHourService.update_man_hour_by_supporter_user_id(
        year=query_params.year,
        month=query_params.month,
        version=query_params.version,
        body=body,
        supporter_user_id=supporter_user_id,
        current_user=current_user,
    )


@router.get(
    "/man-hours/summary/type",
    tags=["ManHour"],
    description="月次工数分類別工数一覧（月次工数集計表）を取得します。",
    response_model=GetSummaryManHourTypeResponse,
    status_code=api_status.HTTP_200_OK,
)
def get_summary_man_hour_type(
    query_params: YaerMonthQuery = Depends(),
    current_user=Depends(
        get_current_user_factory(
            accessible_roles={
                UserRoleType.SYSTEM_ADMIN.key,
                UserRoleType.SALES_MGR.key,
                UserRoleType.SUPPORTER_MGR.key,
                UserRoleType.MAN_HOUR_OPS.key,
                UserRoleType.BUSINESS_MGR.key,
            }
        )
    ),
) -> GetSummaryManHourTypeResponse:
    """Get /api/man-hours/summary/type 月次工数分類別工数一覧取得API

    Args:
        year (int): 対象年
        month (int): 対象月
        current_user (Behavior, optional): 認証済みのユーザー

    Returns:
        GetSummaryManHourTypeResponse: 取得結果
    """

    return ManHourService.get_summary_man_hour_type(
        year=query_params.year,
        month=query_params.month,
        current_user=current_user,
    )


@router.get(
    "/man-hours/summary/supporter",
    tags=["ManHour"],
    description="月次担当者別工数一覧を取得します。 担当者別月次工数集計で利用",
    response_model=GetSummarySupporterManHoursResponse,
    status_code=api_status.HTTP_200_OK,
)
def get_summary_supporter_man_hours(
    query_params: GetSummarySupporterManHoursQuery = Depends(),
    current_user=Depends(
        get_current_user_factory(
            accessible_roles={
                UserRoleType.SYSTEM_ADMIN.key,
                UserRoleType.SALES_MGR.key,
                UserRoleType.SUPPORTER_MGR.key,
                UserRoleType.MAN_HOUR_OPS.key,
                UserRoleType.BUSINESS_MGR.key,
            }
        )
    ),
) -> GetSummaryManHourTypeResponse:
    """Get /api/man-hours/summary/supporter 月次担当者別工数一覧取得API

    Args:
        year (int): 対象年
        month (int): 対象月
        current_user (Behavior, optional): 認証済みのユーザー

    Returns:
        GetSummarySupporterManHoursResponse: 取得結果
    """

    return ManHourService.get_summary_supporter_man_hours(
        year=query_params.year,
        month=query_params.month,
        supporter_organization_id=query_params.supporter_organization_id,
        current_user=current_user,
    )
